contrib/message-capture/message-capture-parser.py

```diff
@@ -0,0 +1,198 @@
+
+            # Determine message type
+            if msgtype not in MESSAGEMAP:
+                # For now just skip unrecognized messages
+                f_in.seek(length, os.SEEK_CUR)
+                print(f"Skipping unrecognized message type {msgtype}", file=sys.stderr)
+                continue
+            msg_ser = BytesIO(f_in.read(length))
+            msg = MESSAGEMAP[msgtype]()
+
+            # Deserialize the message
+            try:
+                msg.deserialize(msg_ser)
+            except KeyboardInterrupt:
+                raise
+            except Exception:
+                # Message was bad, skip the rest of the file
+                print(f"Found broken message, skipping remainder of file: {path}", file=sys.stderr)
+                break
+
+            # Convert message to dictionary and add it to the messages dict
+            msg_dict = {}
+            msg_dict["msgtype"] = getattr(msg, "msgtype", None).decode()
+            msg_dict["direction"] = "recv" if recv else "sent"
+            msg_dict["time"] = time
+            msg_dict["size"] = length   # "size" is less readable here, but more readable in the output
+            if length:
+                msg_dict["body"] = to_jsonable(msg)
+            messages.append(msg_dict)
+
+        if progress_bar:
+            # Update the progress bar to the end of the current file
+            # in case we exited the loop early
+            f_in.seek(0, os.SEEK_END)   # Go to end of file
+            diff = f_in.tell() - bytes_read - 1
+            progress_bar.update(diff)
+
+
+def main():
+    """Main"""
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        epilog="EXAMPLE \n\t{0} -o out.json <data-dir>/message_capture/**/*.dat".format(sys.argv[0]),
+        formatter_class=argparse.RawTextHelpFormatter)
+    parser.add_argument(
+        "capturepaths",
+        nargs='+',
+        help="binary message capture files to parse.")
+    parser.add_argument(
+        "-o", "--output",
+        help="output file.  If unset print to stdout")
+    parser.add_argument(
+        "-n", "--no-progress-bar",
+        action='store_true',
+        help="disable the progress bar.  Automatically set if the output is not a terminal")
+    args = parser.parse_args()
+    capturepaths = [Path.cwd() / Path(capturepath) for capturepath in args.capturepaths]
+    output = Path.cwd() / Path(args.output) if args.output else False
+    use_progress_bar = (not args.no_progress_bar) and sys.stdout.isatty()
+
+    messages = []   # type: List[Any]
+    if use_progress_bar:
+        total_size = sum(capture.stat().st_size for capture in capturepaths)
+        progress_bar = ProgressBar(total_size)
+    else:
+        progress_bar = None
+
+    for capture in capturepaths:
+        process_file(str(capture), messages, "recv" in capture.stem, progress_bar)
+
+    messages.sort(key=lambda msg: msg['time'])
```

Comment:
Yes, definitely a good suggestion. I agree this can wait for a future PR.

See the use of `heapq.merge()` in `test/functional/combine_logs.py` for one way to do this (although I never enhanced that to support streaming since there wasn't an immediate need).

Reply:
I agree on all accounts: this is a naive solution but I'll leave an improvement for a future PR.